src/Tracing.jl

```diff
@@ -201,9 +125,88 @@ function traced_type(::Type{T}, seen, mode, track_numbers) where {T}
     throw(NoFieldMatchError(T, TT2))
 end
 
-function traced_type(
-    ::Type{<:ConcreteRNumber{T}}, seen, ::Val{mode}, track_numbers
-) where {T,mode}
+for T in (DataType, Module, Nothing, Symbol, AbstractChar, AbstractString, AbstractFloat, Integer, RNumber)
+    @eval Base.@nospecializeinfer function traced_type(@nospecialize(T::Type{<:$T}), seen, mode, track_numbers)
+        return T
+    end
+end
+
+Base.@nospecializeinfer function traced_type(
+    @nospecialize(T::Type{<:ReactantPrimitive}), seen, mode::Val{Mode}, track_numbers
+)
+    if Mode == ArrayToConcrete && any(Base.Fix1(<:, T), track_numbers)
+        return ConcreteRNumber{T}
+    end
+    return T
+end
+
+Base.@nospecializeinfer function traced_type(
+    @nospecialize(C::Type{<:Complex}), seen::ST, mode::Val{Mode}, track_numbers::TN
```

Comment:
i think we would benefit here too because it's just passed

```suggestion
    @nospecialize(C::Type{<:Complex}), seen::ST, @nospecialize(mode::Val{Mode}), track_numbers::TN
```